### ISSUE TITLE
test model performance on validation dataset

### ISSUE THREAD

**Issue by lijinmaylee** [Rambatino/CHAID#23]
Hi, thank you very much for the implementation!

I plan to split my dataset into development and validation set, then I would like to build the CHAID tree on development set, and test its performance on validation set.

It would be really nice if there could be a function that outputs the rules of segmentation and/or applies the rules (i.e. does model prediction) on the validation set.

Thanks a lot for your help!

**Comment by Rambatino** [comment]
Hi @lijinmaylee, could you please elaborate on exactly what you mean. What are these rules of segmentation?
Also, if you want to build the feature via a Pull Request we'd be happy to review it.

**Comment by lijinmaylee** [comment]
Hi, thanks for your reply!

What I mean by outputting the decision tree rules is similar to what is described in this thread [http://stackoverflow.com/questions/28428508/chaid-regression-tree-to-table-conversion-in-r].

And by saying test its performance on validation set, what I expect is something similar to the predict function in R, which is, for a new data point, given the decision tree built on the development set, I could know which node it falls into, and what the model prediction for its class label is.

Thanks for your consideration!

**Comment by Rambatino** [comment]
Is it something like this: 

```    /* Node 31 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 = 0)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_2 = 0)  AND  (S16_Top2_6 != 0)
THEN
Node = 31
Prediction = 0
Probability = 0.697674

/* Node 32 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 = 0)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_2 = 0)  AND  (S16_Top2_6 = 0)
THEN
Node = 32
Prediction = 1
Probability = 0.586538

/* Node 33 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 = 0)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_2 != 0)  AND  (S16_Top2_17 != 0)
THEN
Node = 33
Prediction = 0
Probability = 0.920592

/* Node 34 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 = 0)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_2 != 0)  AND  (S16_Top2_17 = 0)
THEN
Node = 34
Prediction = 0
Probability = 0.683333

/* Node 35 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 = 0)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_22 != 0)  AND  (S16_Top2_17 != 0)
THEN
Node = 35
Prediction = 0
Probability = 0.671795

/* Node 36 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 = 0)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_22 != 0)  AND  (S16_Top2_17 = 0)
THEN
Node = 36
Prediction = 1
Probability = 0.625000

/* Node 37 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 = 0)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_22 = 0)  AND  (S16_Top2_2 = 0)
THEN
Node = 37
Prediction = 1
Probability = 0.940299

/* Node 38 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 = 0)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_22 = 0)  AND  (S16_Top2_2 != 0)
THEN
Node = 38
Prediction = 1
Probability = 0.582418

/* Node 39 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 != 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_2 = 0)
THEN
Node = 39
Prediction = 0
Probability = 0.937018

/* Node 40 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 != 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_2 != 0)
THEN
Node = 40
Prediction = 0
Probability = 0.992855

/* Node 41 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 != 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_22 != 0)
THEN
Node = 41
Prediction = 0
Probability = 0.921053

/* Node 42 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 != 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_22 = 0)
THEN
Node = 42
Prediction = 0
Probability = 0.705479

/* Node 43 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 != 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_22 != 0)  AND  (S16_Top2_24 != 0)
THEN
Node = 43
Prediction = 0
Probability = 0.914956

/* Node 44 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 != 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_22 != 0)  AND  (S16_Top2_24 = 0)
THEN
Node = 44
Prediction = 0
Probability = 0.656566

/* Node 45 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 != 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_22 = 0)  AND  (S16_Top2_2 = 0)
THEN
Node = 45
Prediction = 1
Probability = 0.719298

/* Node 46 */.
IF (S16_Top2_27 != 0)  AND  (S16_Top2_21 != 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_22 = 0)  AND  (S16_Top2_2 != 0)
THEN
Node = 46
Prediction = 0
Probability = 0.725490

/* Node 47 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_21 != 1)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_2 = 0)
THEN
Node = 47
Prediction = 1
Probability = 0.704225

/* Node 48 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_21 != 1)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_2 != 0)
THEN
Node = 48
Prediction = 0
Probability = 0.585062

/* Node 49 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_21 != 1)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_2 = 0)
THEN
Node = 49
Prediction = 1
Probability = 0.986301

/* Node 50 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_21 != 1)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_2 != 0)
THEN
Node = 50
Prediction = 1
Probability = 0.690323

/* Node 51 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_21 = 1)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_34 = 0)
THEN
Node = 51
Prediction = 0
Probability = 0.754237

/* Node 52 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_21 = 1)  AND  (S16_Top2_24 != 0)  AND  (S16_Top2_34 != 0)
THEN
Node = 52
Prediction = 0
Probability = 0.931034

/* Node 53 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_21 = 1)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_22 != 0)
THEN
Node = 53
Prediction = 0
Probability = 0.682927

/* Node 54 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 != 0)  AND  (S16_Top2_21 = 1)  AND  (S16_Top2_24 = 0)  AND  (S16_Top2_22 = 0)
THEN
Node = 54
Prediction = 1
Probability = 0.648148

/* Node 55 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_24 = 1)  AND  (S16_Top2_21 != 1)  AND  (S16_Top2_12 = 0)
THEN
Node = 55
Prediction = 1
Probability = 0.984848

/* Node 56 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_24 = 1)  AND  (S16_Top2_21 != 1)  AND  (S16_Top2_12 != 0)
THEN
Node = 56
Prediction = 1
Probability = 0.623188

/* Node 57 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_24 = 1)  AND  (S16_Top2_21 = 1)  AND  (S16_Top2_1 != 0)
THEN
Node = 57
Prediction = 0
Probability = 0.707317

/* Node 58 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_24 = 1)  AND  (S16_Top2_21 = 1)  AND  (S16_Top2_1 = 0)
THEN
Node = 58
Prediction = 1
Probability = 0.641026

/* Node 59 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_24 != 1)  AND  (S16_Top2_22 = 1)  AND  (S16_Top2_21 != 1)
THEN
Node = 59
Prediction = 1
Probability = 0.924528

/* Node 60 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_24 != 1)  AND  (S16_Top2_22 = 1)  AND  (S16_Top2_21 = 1)
THEN
Node = 60
Prediction = 1
Probability = 0.727273

/* Node 61 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_24 != 1)  AND  (S16_Top2_22 != 1)  AND  (S16_Top2_2 != 1)
THEN
Node = 61
Prediction = 1
Probability = 1.000000

/* Node 62 */.
IF (S16_Top2_27 = 0)  AND  (S16_Top2_17 = 0)  AND  (S16_Top2_24 != 1)  AND  (S16_Top2_22 != 1)  AND  (S16_Top2_2 = 1)
THEN
Node = 62
Prediction = 1
Probability = 0.941176```

I'm having a hard way of pinning it down exactly what you mean, even with that link.

If that isn't the kind of thing you want, could you please be very specific in your request and we'll try and build it.

**Comment by lijinmaylee** [comment]
Hi, 

That is exactly what I meant.

Thanks so much!

**Comment by xulaus** [comment]
Hi @lijinmaylee, did you want this a text output, or some sort of object that will do the classification for you?

**Comment by Rambatino** [comment]
It shouldn't be too hard to build, and would be good for debugging. I'd prefer a more structured output though

**Comment by lijinmaylee** [comment]
hi @xulaus , an object would be preferred. 

Thanks a lot for your effort!

**Comment by xulaus** [comment]
Still some argument here about what you mean. Would a pandas data frame fit your use case? If not could you provide some example code of how you would like the API to look?

**Comment by lijinmaylee** [comment]
Hi, 

I would expect that the test/validation dataset will have exactly the same structure as the development dataset - both are pandas dataframes with same columns.

Hope that this clarifies the confusion. Thanks for your consideration.

**Comment by Tizpaz** [comment]
Hello,

Could you please help me, how I can find performance (accuracy) of a CHIAD model? I use cross_validation and split the data to 80% for train and 20% for test. After training the CHAID decision tree, now I want to know how I can measure the accuracy for 20% of test data:

X_train, X_test, y_train, y_test = cross_validation.train_test_split(X,y,test_size=0.2)

tree = CHAID(X_train,y_train,max_depth=20,split_titles=header).to_tree().show()

?? (* how to measure the accuracy of model for 20% of test data *)

Thanks,
Saeid

**Comment by Rambatino** [comment]
Are these the same issues? We're most likely going to be building a way of returning the decision rules of the tree. 

How are you defining 'accuracy' @Tizpaz ? Can you point us to some resources and we can look into building this feature

**Comment by Tizpaz** [comment]
Maybe this is another issue. In Machine Learning, after training a classifier (like CHAID) for example, with 80% of data set, we need to measure the accuracy of the classifier using the remain 20%. This is usually done be cross validation function which splits the data to two train and test data set. Now, the accuracy is the rate of correct prediction of class label for the test set: Accuracy = (True Positive + True Negative) / (True Positive + True Negative + False Positive + False Negative) for binary class label. Please refer to: https://en.wikipedia.org/wiki/Evaluation_of_binary_classifiers for more information.

**Comment by Rambatino** [comment]
Okay, so I completely get what you're asking for now. 

We have two related, but separate concepts here.

This issue outlined, originally here, is primarily dealing with stopping rules; something that is in progress of be being built (I think there may be a pull request for it) which will form version 0.4.0.

What you are asking for is to determine the accuracy of the model. You have a subset and you essentially train the model on it, you then apply that same logic which derived the respondents' terminal nodes in the first set (e.g. people who are male, 20-25 and blonde fall into this terminal node (which is what this issue here deals with)) and then you look for at the same set of characteristics for individuals in the second data set and work out the frequency by which those stopping rules _still apply_.

Let's carry on this conversation in another issue (i'll open one now) and describe how I think this will work.

**Comment by Rambatino** [comment]
There is a new PR (which will need some work https://github.com/Rambatino/CHAID/pull/36, which will produce the stopping rules.. or at the very least this method _should be_ a precursor to that)

Also, we do have a 'risk' calculation, which looks at the observed dependent choice, compared to the predicted (as defined by the highest choice that row fell into) and calculates how well the model predicts the observed value

**Comment by xulaus** [comment]
Fixed with PR #47, will be in version 3.0.0
